### PR TITLE
br, lightning: move *batch* split region logic into common package

### PR DESCRIPTION
**Pull request by lance6716**
### What problem does this PR solve?

Issue Number: ref #51533

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
   BR https://do.pingcap.net/jenkins/blue/organizations/jenkins/pingcap%2Ftidb%2Fpull_br_integration_test/detail/pull_br_integration_test/99/pipeline
   lightning https://do.pingcap.net/jenkins/blue/organizations/jenkins/pingcap%2Ftidb%2Fpull_lightning_integration_test/detail/pull_lightning_integration_test/152/pipeline
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by D3Hunter — COMMENTED** [review]
8/11

File: br/pkg/restore/split/client.go [hunk 1]
```diff
@@ -555,39 +565,67 @@ func (c *pdClient) hasHealthyRegion(ctx context.Context, regionID uint64) (bool,
...
+			err = c.scatterRegions(ctx, newRegionsOfBatch)
+			if err != nil {
+				brlog.FromContext(ctx).Warn(
+					"scatter regions failed, will continue anyway",
+					zap.Error(err),
+				)
+			}
+
+			// the region with the max start key is the region need to be further split, it
+			// can be the leftmost region or the rightmost region.
```
Comment (lance6716) [inline br/pkg/restore/split/client.go:610]: by checking the code of TiKV, seems it's always the rightmost region. The region with original region ID can be the first or the last region.

Raftstore v2

https://github.com/tikv/tikv/blob/52947275104ccfdafb6dc1a2e580bdbecf6cc3f4/components/raftstore-v2/src/operation/command/admin/split.rs#L512-L535

v1

https://github.com/tikv/tikv/blob/52947275104ccfdafb6dc1a2e580bdbecf6cc3f4/components/raftstore/src/store/fsm/apply.rs#L2588-L2631

I'll remove this check and directly use the last region

File: pkg/lightning/backend/local/localhelper.go [hunk 2]
```diff
@@ -169,73 +166,29 @@ func (local *Backend) SplitAndScatterRegionByRanges(
...
+						if common.IsContextCanceledError(err2) {
+							return err2
 						}
-						endIdx++
+						log.FromContext(ctx).Warn("split regions",
+							log.ShortError(err2),
+							zap.Int("retry time", i),
+							zap.Uint64("region_id", region.Region.Id))
+						// TODO(lance6716): now the retryKeys can not be shrank to the middle of the keys
+						// and it will retry all keys of the region. Will fix it in future PR.
```
Comment (D3Hunter) [inline pkg/lightning/backend/local/localhelper.go:181]: also need to pick to 8.1 if the pr is not merged before 4.9

**Review by D3Hunter — APPROVED** [review]

File: br/pkg/restore/split/client.go [hunk 3]
```diff
@@ -555,39 +572,68 @@ func (c *pdClient) hasHealthyRegion(ctx context.Context, regionID uint64) (bool,
...
+				brlog.FromContext(ctx).Warn(
+					"scatter regions failed, will continue anyway",
+					zap.Error(err),
+				)
+			}
+
+			// the region with the max start key is the region need to be further split,
+			// depending on the origin region is the first region or last region, we need to
+			// compare the origin region and the last one of new regions.
+			lastNewRegion := newRegionsOfBatch[len(newRegionsOfBatch)-1]
```
Comment (Leavrth) [inline br/pkg/restore/split/client.go:619]: it will panic if the `newRegionsOfBatch` is nil when `batchSplitRegionsWithOrigin` returns error.

**Review by Leavrth — COMMENTED** [review]
LGTM

File: br/pkg/restore/split/client.go (hunk 3 above)
Comment (lance6716) [reply br/pkg/restore/split/client.go:619]: we have checked the error in line 595, that error check will return

**Review by Leavrth — APPROVED** [review]

**Pull request merged**